jdaviz/core/template_mixin.py

```diff
@@ -856,3 +832,163 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # NOTE: cannot be named self.data or will conflict with existing self.data traitlet!
         self.dataset = DatasetSelect(self, 'dataset_items', 'dataset_selected')
+
+
+class AddResults(BasePluginComponent):
+    """
+    Traitlets (in the object, custom traitlets in the plugin):
+
+    * ``label`` (string: user-provided label for the results data-entry.  If ``label_auto``, changes
+        to ``label_default`` will update ``label``.  Otherwise, changes to ``label`` will set
+        ``label_auto`` to False.)
+    * ``label_default`` (string: plugin-determined default label that will be synced to ``label``
+        if/when ``label_auto`` is set to True)
+    * ``label_auto`` (bool: whether to sync ``label_default`` to ``label``)
+    * ``label_invalid_msg`` (string: validation string for the current value of ``label``.  If
+        not an empty string, calls to ``add_results_from_plugin`` will raise an error.)
+    * ``label_overwrite`` (bool: whether the value of ``label`` already exists for a data-entry
+        from the same plugin)
+    * ``add_to_viewer_items`` (list of dicts: see ``ViewerSelect``)
+    * ``add_to_viewer_selected`` (string: name of the viewer to add the results,
+        see ``ViewerSelect``)
+
+
+    Methods:
+
+    * ``add_results_from_plugin``
+
+
+    Example template::
+
+      <plugin-add-results
+        :label.sync="results_label"
+        :label_default="results_label_default"
+        :label_auto.sync="results_label_auto"
+        :label_invalid_msg="results_label_invalid_msg"
+        :label_overwrite="results_label_overwrite"
+        label_hint="Label for the smoothed data"
+        :add_to_viewer_items="add_to_viewer_items"
+        :add_to_viewer_selected.sync="add_to_viewer_selected"
+        action_label="Apply"
+        action_tooltip="Apply the action to the data"
+        @action-click="apply"
+      ></plugin-add-results>
+
+    """
+    def __init__(self, plugin, label, label_default, label_auto,
+                 label_invalid_msg, label_overwrite,
+                 add_to_viewer_items, add_to_viewer_selected):
+        super().__init__(plugin, label=label,
+                         label_default=label_default, label_auto=label_auto,
+                         label_invalid_msg=label_invalid_msg, label_overwrite=label_overwrite,
+                         add_to_viewer_items=add_to_viewer_items,
+                         add_to_viewer_selected=add_to_viewer_selected)
+
+        # DataCollectionAdd/Delete are fired even if remain unchecked in all viewers
+        self.hub.subscribe(self, DataCollectionAddMessage,
+                           handler=lambda _: self._on_label_changed())
+        self.hub.subscribe(self, DataCollectionDeleteMessage,
+                           handler=lambda _: self._on_label_changed())
```

Comment:
lambda functions here are intentional as `_on_label_changed` is expecting an event dictionary, not a message object.